src/zenml/config/compiler.py

```diff
@@ -389,9 +377,12 @@ def _compile_step(
         Returns:
             The compiled step.
         """
-        step_spec = self._get_step_spec(
-            pipeline_parameter_name=pipeline_parameter_name, step=step
-        )
+        # Copy the invocation (including its referenced step) before we apply
+        # the step configuration which is exclusive to this invocation.
+        invocation = copy.deepcopy(invocation)
```

Comment:
Yes that is necessary, because multiple invocations can reference a step which might later get configured. The invocations need access to the step instance so they can get the latest configuration before the compilation.